File: src/memento/graph_debug/refresh.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from pathlib import Path

from memento.config import GraphExplorerConfig
from memento.derived.embeddings_worker import SemanticEmbeddingRefreshWorker
from memento.graph_debug.snapshot import GraphSnapshotError, GraphSnapshotService


@dataclass(frozen=True, slots=True)
class GraphEmbeddingRefreshState:
    available: bool
    running: bool
    pending: bool
    last_error: str | None
    last_scope: str | None
    queued_paths: int
    repository_revision: str | None


class GraphEmbeddingRefreshCoordinator:
    def __init__(
        self,
        config: GraphExplorerConfig,
        *,
        repository_root: Path,
        snapshot_service: GraphSnapshotService,
        worker: SemanticEmbeddingRefreshWorker | None,
    ) -> None:
        self._config = config
        self._repository_root = repository_root
        self._snapshot_service = snapshot_service
        self._worker = worker
        self._last_scope: str | None = None
        self._queued_paths = 0
        self._repository_revision: str | None = None

    def enqueue(
        self,
        *,
        scope: str,
        concept_ids: tuple[str, ...] = (),
        confirm_full: bool = False,
    ) -> GraphEmbeddingRefreshState:
        if self._worker is None:
            raise GraphSnapshotError("semantic embedding refresh is unavailable")
        overview = self._snapshot_service.overview()
        revision = overview.revisions.repository
        paths: tuple[str, ...] | None
        if scope == "full":
            if not confirm_full:
                raise GraphSnapshotError("full embedding refresh requires confirmation")
            paths = None
            queued = overview.metrics.memory_count
        elif scope in {"selected", "visible"}:
            if not concept_ids:
                raise GraphSnapshotError(f"{scope} embedding refresh requires concept ids")
            unique = tuple(sorted(dict.fromkeys(concept_ids)))
            if len(unique) > self._config.refresh_max_paths:
                raise GraphSnapshotError("embedding refresh exceeds configured path limit")
            paths = tuple(self._snapshot_service.paths_for_ids(unique))
            if len(paths) != len(unique):
                raise GraphSnapshotError("embedding refresh includes unknown memories")
            queued = len(paths)
        else:
            raise GraphSnapshotError("unsupported embedding refresh scope")
        if not self._worker.enqueue(
            self._repository_root,
            revision,
            paths=paths,
        ):
            raise GraphSnapshotError("embedding refresh worker is closed")
        self._last_scope = scope
        self._queued_paths = queued
        self._repository_revision = revision
        return self.state()

    def state(self) -> GraphEmbeddingRefreshState:
        if self._worker is None:
            return GraphEmbeddingRefreshState(
                available=False,
                running=False,
                pending=False,
                last_error=None,
                last_scope=self._last_scope,
                queued_paths=self._queued_paths,
                repository_revision=self._repository_revision,
            )
        state = self._worker.state()
        return GraphEmbeddingRefreshState(
            available=True,
            running=state.running,
            pending=state.pending,
            last_error=state.last_error,
            last_scope=self._last_scope,
            queued_paths=self._queued_paths,
            repository_revision=self._repository_revision,
        )

    def state_dict(self) -> dict[str, object]:
        return asdict(self.state())

File: go/graphdebug/refresh.go
package graphdebug

import (
	"context"
	"sort"
	"sync"

	"github.com/rcarmo/memento/go/access"
)

type RefreshWorkerState struct {
	Alive, Running, Pending             bool
	LastError, PauseReason, CurrentPath *string
	Completed                           int
}
type RefreshWorker interface {
	Enqueue(repositoryRoot, revision string, paths []string, full bool) bool
	State() RefreshWorkerState
}
type RefreshState struct {
	Available          bool    `json:"available"`
	Running            bool    `json:"running"`
	Pending            bool    `json:"pending"`
	LastError          *string `json:"last_error"`
	LastScope          *string `json:"last_scope"`
	QueuedPaths        int     `json:"queued_paths"`
	RepositoryRevision *string `json:"repository_revision"`
	PauseReason        *string `json:"pause_reason"`
	CurrentPath        *string `json:"current_path"`
	Completed          int     `json:"completed"`
	Alive              bool    `json:"alive"`
}
type RefreshSnapshots interface {
	Overview(context.Context, *access.EffectivePolicy, OverviewOptions) (Overview, error)
	PathsForIDs(context.Context, []string) ([]string, error)
}

type RefreshCoordinator struct {
	Service                                     RefreshSnapshots
	Worker                                      RefreshWorker
	RepositoryRoot                              string
	RefreshMaxPaths, DirectNodeLimit, EdgeLimit int
	lastScope                                   *string
	queuedPaths                                 int
	repositoryRevision                          *string
	mu                                          sync.RWMutex
}

func (c *RefreshCoordinator) State() RefreshState {
	c.mu.RLock()
	lastScope, queued, revision := copyText(c.lastScope), c.queuedPaths, copyText(c.repositoryRevision)
	c.mu.RUnlock()
	if c.Worker == nil {
		return RefreshState{LastScope: lastScope, QueuedPaths: queued, RepositoryRevision: revision}
	}
	state := c.Worker.State()
	return RefreshState{Available: state.Alive, Alive: state.Alive, Running: state.Running, Pending: state.Pending, LastError: copyText(state.LastError), LastScope: lastScope, QueuedPaths: queued, RepositoryRevision: revision, PauseReason: copyText(state.PauseReason), CurrentPath: copyText(state.CurrentPath), Completed: state.Completed}
}
func copyText(value *string) *string {
	if value == nil {
		return nil
	}
	copy := *value
	return &copy
}
func (c *RefreshCoordinator) Enqueue(ctx context.Context, scope string, conceptIDs []string, confirmFull bool) (RefreshState, error) {
	if c.Worker == nil {
		return RefreshState{}, &SnapshotError{"semantic embedding refresh is unavailable"}
	}
	overview, err := c.Service.Overview(ctx, nil, OverviewOptions{DirectNodeLimit: c.DirectNodeLimit, EdgeLimit: c.EdgeLimit, RefreshMaxPaths: c.RefreshMaxPaths})
	if err != nil {
		return RefreshState{}, err
	}
	revision := overview.Revisions.Repository
	var paths []string
	full := false
	queued := 0
	switch scope {
	case "full":
		if !confirmFull {
			return RefreshState{}, &SnapshotError{"full embedding refresh requires confirmation"}
		}
		full = true
		queued = overview.Metrics.MemoryCount
	case "selected", "visible":
		if len(conceptIDs) == 0 {
			return RefreshState{}, &SnapshotError{scope + " embedding refresh requires concept ids"}
		}
		uniqueMap := map[string]bool{}
		for _, id := range conceptIDs {
			uniqueMap[id] = true
		}
		unique := make([]string, 0, len(uniqueMap))
		for id := range uniqueMap {
			unique = append(unique, id)
		}
		sort.Strings(unique)
		if len(unique) > c.RefreshMaxPaths {
			return RefreshState{}, &SnapshotError{"embedding refresh exceeds configured path limit"}
		}
		paths, err = c.Service.PathsForIDs(ctx, unique)
		if err != nil {
			return RefreshState{}, err
		}
		if len(paths) != len(unique) {
			return RefreshState{}, &SnapshotError{"embedding refresh includes unknown memories"}
		}
		queued = len(paths)
	default:
		return RefreshState{}, &SnapshotError{"unsupported embedding refresh scope"}
	}
	if !c.Worker.Enqueue(c.RepositoryRoot, revision, paths, full) {
		return RefreshState{}, &SnapshotError{"embedding refresh worker is closed or stopped"}
	}
	scopeCopy, revisionCopy := scope, revision
	c.mu.Lock()
	c.lastScope = &scopeCopy
	c.queuedPaths = queued
	c.repositoryRevision = &revisionCopy
	c.mu.Unlock()
	return c.State(), nil
}
